# Prerequisitos con Python - Parte 4

Vimos varias funciones que python tiene built in.  Ahora veremos funciones que podemos construir nosotros.  Las funciones las utilizamos para cuando tenemos un código que se repite o queremos realizar la misma accion repetidas veces.

Supongamos que tenemos un sistema de IA que queremos desarrollar un sistema.  El sistema tiene un modulo que hace lo siguiente.
- Training o Modelado
- Testing o Validacion
- Produccion
- Reportes
- Logging

Para desarrollar más comodamente nos dedicamos entonces a crear funciones para cada parte pues algunas secciones se repiten a lo largo del desarrollo, probablemente logging lo requieran en todas partes del sistema.

En resumen, las funciones nos ayudan a simplificar el código y a optimizarlo.

### Nuestra primera función

Las funciones tienen argumentos de entrada, salida o pueden que no tengan argumentos alguno ni de entrada o ni de salida.  Para la siguiente función realizaremos aquella que nos permita calcular la distancia en función de una velocidad inicial, el tiempo y la aceleración.

In [1]:
def distancia(vo, dt, a):
    d = vo*dt + 0.5*a*dt**2
    return d

vo = 2
t = 0.5
a = 1.2
d = distancia(vo, t, a)
print(d)

1.15


Podemos tambien llamar a la funcion ingresando los argumentos directamente

In [3]:
d = distancia(2, 0.5, 1.2)
print(d)

1.15


¿Qué sucede aqui?.  

Tenemos entonces que la funcion siempre empieza por la palabra 'def', lo que nos dice que es una funcion, y dicha funcion siempre debe terminar con ':'.  

El termino 'distancia' es el nombre de la función y debe ser sin separaciones, es decir, digamos que se llame ahora calcular distancia en vez de distancia, el nombre correcto de la función sería 'calcular_distancia'.

Los valores entre '()' son los argumentos de la función, estos son separados por ',', si no necesitamos argumentos simplemente podemos definirlas como por ejemplo:

    def printing():
        print('ppppp')

Dentro de la función (cuerpo) podemos definir y usar las variables.  Las variables que se declaren dentro solamente se pueden utilizar internamente, pero hablaremos de esto luego, debajo veremos que la funcion tiene un valor kp que es solo accesible en la función.

    def printing():
        strx = 'pppp'
        print(strx)

El valor de retorno es aquel que se da al final de la función, especificado por 'return', si una función no tiene retorno se debe realiar sin este token (return) asi como las funciones anteriores

    def mimic(a)
       return a
       
Por defecto, las funciones que no tienen ningun tipo de retorno su valor de regreso es None.  Veamos si es cierto.

In [5]:
retorno = print(' ')

print(retorno)

 
None


### Ahora ud.


Realizar una funcion que calcule el área de una elipse.

In [1]:
from math import pi
def elipse(a,b):
    area_elipse=pi*a*b
    return area_elipse

a = 10
b = 5
area_elipse=elipse(a,b)

print(f'{area_elipse:.{7}f}')

157.0796327


Respuesta esperada:

157.0796327

Realice una función que dada el tiempo como string ("465:50:00") calcule a cuantos segundos equivalen

In [2]:
t = "520:59:30"
def segundos(t):
    HORAS, MINUTOS, SEGUNDOS = t.split(':')
    S=int(HORAS) * 3600 + int(MINUTOS) * 60 + int(SEGUNDOS)
    return S
S=segundos(t)
print(S)


1875570


Respuesta esperada:

1875570

#### Alcance de Variables

Cuando hablamos de alcance de variables se refiere hasta donde puede ser utilizada esta variable (donde puede ser invocada) en el transcurso de un programa.

Si la variable es creada dentro de una función, esta solo tiene alcance dentro de la misma, no puede ser llamada fuera de ella.  Veamos el ejemplo.

In [17]:
def area_rectangulo(base, altura):
    area = base * altura
    return area

def area_triangulo(base, altura):
    area = 0.5 * base * altura
    return area

b = 2
h = 3 
area1 = area_rectangulo(b, h)
area2 = area_triangulo(b, h)
print(area1, area2)

6 3.0


Veamos un ejemplo de una variable fuera de la funcion.  Aqui, esta variable puede ser accedida en cualquier función o fuera de ella porque su alcance es mas 'global'.

In [20]:
val = 15

def print_val():
    print(val)
    
print_val()
val

15


15

#### Ahora ud 

Del siguiente código cual es el error y porqué:

In [4]:
def area_circulo(r):
    pi = 3.141592654
    print(pi*r**2)

area_circulo(4)
pi_medio = pi/2

50.265482464


RESPUESTA: PI SOLO SE PUEDE UTILIZAR DENTRO DE AREA_CIRCULO POR ESO NO LLEGA A RECONOCERLA.

#### Documentación de código

Es importante cuando se realiza código documentarlo.  De esta forma se puede interpretar rápido que se realizó y ayuda a otros desarrolladores.

Les puedo dar la importancia en un ejemplo.  Imaginese un libro de cálculo con todos los problemas resueltos de manera correcta pero sin un capítulo escrito, el libro, aunque sea muy bueno cuesta entenderlo para un novato.

En la documentación, es bueno tener ciertas reglas, como por ejemplo, no nombrar variables sin sentido, hacerla lo más humanamente entendible.

Veamos una manera correcta de documentar.

In [25]:
def area_trapecio(a, b, h):
    area = (a + b)*h/2 # a y b, base y lado superior respectivamente; h es la altura
    return area


# calculando el area de diferentes trapezoides
print(area_trapecio(4, 2, 1))
print(area_trapecio(3, 8, 8))
print(area_trapecio(1, 1, 0.2))
print(area_trapecio(7, 3, 0.5))


# escribiendo lo mismo pero sin funciones
a, b, h = 4, 2, 1
area = (a + b)*h/2
print(area)

a, b, h = 3, 8, 8
area = (a + b)*h/2
print(area)

a, b, h = 1, 1, 0.2
area = (a + b)*h/2
print(area)

a, b, h = 7, 3, 0.5
area = (a + b)*h/2
print(area)

3.0
44.0
0.2
2.5
3.0
44.0
0.2
2.5


Podemos tambien utilizar docstring para ayudar a la documentación.  Veamos como se pueden usar para documentación con la misma función anterior

In [26]:
def area_trapecio(a, b, h):
    """
    CALCULA EL AREA DEL TRAPEZOIDE DATA BASE, ALTURA Y PARTE SUPERIOR
                   a
                 ______       _
               /       \      |
              /         \     h
             /___________\   _|_
                   b
    
    ENTRADAS:
    a:  parte superior del trapezoide
    b:  parte inferior (base) del trapezoide
    h:  altura del trapezoide
    
    RETORNA:
    area_trapecio: el area del trapecio dada por la ecuación.
    """
    area = (a + b)*h/2 # a y b, base y lado superior respectivamente; h es la altura
    return area

#### Ahora ud

Escriba un docstring para realizar la función de cálculo del área de un paralelogramo

In [5]:
def area_paralelogramo(b,h):
    """
    CALCULAR EL AREA DE UN PARALELOGRAMO
    Existen 4 tipos de paralelogramos: cuadrados, rectangulos, rombos y romboides
     ________   ___     ______________   ___       ________            ___________    ___
    |        |   |     |              |   |        \      |\         /            /    |
    |        |   h     |              |   h         \     h \       /            /     h
    |________|  _|_    |______________|  _|_         \____|__\     /____________/     _|_
         b                      b                         b               b
        

    ENTRADAS:
    b: base del paralelogramo (parte inferior)
    h: altura del paralelogramo 

    RETORNA:
    area_paralelogramo: area del paralelogramo dada por la ecuacion.
    """
    area=b*h
    return area

#### Expresiones Lambda

En python se pueden utilizar funciones que no tienen un nombre (son anonimas), estas expresiones se llaman funciones lambda.

Comparemos una funcion y una expresión lambda viendo un ejemplo:

In [27]:
def suma(a, b):
    return a+b

print(suma(3,2))

5


In [30]:
suma = lambda x,y: x + y
print(suma(3,2))

5


He aquí una evaluación de una función lambda, luego de la función lambda tenemos los argumentos, idem que en una función, seguido el ':' y finalmente lo que vamos a ejecutar dentro de esta función

In [33]:
n2 = lambda n: n**2
print(n2(5))

25


Lo anterior seria el equivalente a una función lambda con un solo argumento

### Intentelo ud.

Utilizaremos ahora map con funciones lambda.  map() es una función built-in de python que toma como argumento una función y retorna un iterador que aplica la función de cada elemento al iterable.  Veamos el siguiente ejemplo para una funcion map().

In [46]:
n_list = [list(range(10)), list(range(2, 20)), list(range(5, 40, 2))]

def avg(n_list):
    av = sum(n_list)/len(n_list)
    return av

promedio = list(map(avg, n_list))
promedio, " ", n_list

([4.5, 10.5, 22.0],
 ' ',
 [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
  [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
  [5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39]])

Escriba el mismo procedimiento pero utilizando funciones lamda

In [6]:

n_list = [list(range(10)), list(range(2, 20)), list(range(5, 40, 2))]


prom = list(map(lambda y: sum(y) / float(len(y)), n_list))
prom

[4.5, 10.5, 22.0]

Resultado esperado:

[4.5, 10.5, 22.0]

filter() es una función built-in de python que toma una función y un iterable y retorna un iterador con elementos del itreable de las cuales retorna correcto.  Veamos el siguiente ejemplo

In [47]:
valores = [1, 20, -1, 2, 5, 4, -2, -5, 10]

def impares(v):
    if v%2 != 0:
        return True

vals = list(filter(impares, valores))
print(vals)

[1, -1, 5, -5]


Ahora realice lo mismo utilizando funciones lambda

In [7]:
valores = [1, 20, -1, 2, 5, 4, -2, -5, 10]
vals= list(filter(lambda x: x % 2 != 0, valores))
vals

[1, -1, 5, -5]

Resultado esperado:

[1, -1, 5, -5]

#### Iteradores y Generadores

Las listas son los iteradores más comunes.  Un iterador es un objeto que representa un stream de datos.  Las listas son de este tipo (iterables) pero no son un iterador.

Veremos entonces como crear iteradores por medio de generadores.  Los generadores nos ayudan a crear iteradores a base de funciones.

Veamos un ejemplo.

In [51]:
def generador(x):
    i = 0
    while i < x:
        yield i
        i += 1
        
print(generador(6))

<generator object generador at 0x0000023CBAD7DD60>


In [52]:
for n in generador(6):
    print(n)

0
1
2
3
4
5


#### Práctica

Escribir su propio generador con la función enumerate

In [10]:
frases = ["Practical GUI", "Advanced Programming", "Building Intelligent", "The AI Workshop", "Data Analysis"]

def enumerador(f, i):
    i=1
    for e in frases:
        yield i, e
        i+=1
    
for i, frase in enumerador(frases, 1):
    print("Book {}: {}".format(i, frase))

Book 1: Practical GUI
Book 2: Advanced Programming
Book 3: Building Intelligent
Book 4: The AI Workshop
Book 5: Data Analysis


Respuesta esperada:

    Book 1: Practical Gui
    Book 2: Advanced Programming 
    Book 3: Build Intelligent
    Book 4: The AI Workshop
    Book 5: Data Analysis

### Generadores

Para producir expresiones generadoras se hace de la misma manera que cuando se utiliza una lista de comprensión, pero con '( )'


In [11]:
gen = (n + 4 for n in range(10))
gen

<generator object <genexpr> at 0x00000173DCAAC900>

In [12]:
for g in gen:
    print(g)

4
5
6
7
8
9
10
11
12
13
